machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,247 @@
+	// that gives someone the power to push images to the image repository. By comparison, if the final image pull secret gets exfiltrated,
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// This pull secret will be used on all nodes in the pool. These nodes will need to pull the final OS image and boot into it using rpm-ostree or bootc.
+	// +optional
+	CurrentImagePullSecret *ImageSecretObjectReference `json:"currentImagePullSecret,omitempty"`
+}
+
+type MachineOSImageBuilder struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// +kubebuilder:validation:Enum:=JobImageBuilder
+	// Valid options are: JobImageBuilder
+	// +required
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+}
+
+// MachineOSContainerfile contains all custom content the user wants built into the image
+type MachineOSContainerfile struct {
+	// containerfileArch describes the architecture this containerfile is to be built for.
+	// This arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=ARM64;AMD64;PPC64LE;S390X;AArch64;x86_64;NoArch
+	// +kubebuilder:default:=NoArch
+	// +optional
+	ContainerfileArch ContainerfileArch `json:"containerfileArch"`
+	// content is an embedded Containerfile/Dockerfile that defines the contents to be built into your image.
+	// See https://github.com/containers/common/blob/main/docs/Containerfile.5.md for the spec reference.
+	// for example, this would add the tree package to your hosts:
+	//   FROM configs AS final
+	//   RUN rpm-ostree install tree && \
+	//     ostree container commit
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=4096
+	Content string `json:"content"`
+}
+
+// +enum
+type ContainerfileArch string
+
+const (
+	// describes the arm64 architecture
+	Arm64 ContainerfileArch = "ARM64"
+	// describes the amd64 architecture
+	Amd64 ContainerfileArch = "AMD64"
+	// describes the ppc64le architecture
+	Ppc ContainerfileArch = "PPC64LE"
+	// describes the s390x architecture
+	S390 ContainerfileArch = "S390X"
+	// describes the aarch64 architecture
+	Aarch64 ContainerfileArch = "AArch64"
+	// describes the fx86_64 architecture
+	X86_64 ContainerfileArch = "x86_64"
+	// describes a containerfile that can be applied to any arch
+	NoArch ContainerfileArch = "NoArch"
+)
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+// the MachineOSBuilder pod validates that the user has provided a valid pool
+type MachineConfigPoolReference struct {
+	// name of the MachineConfigPool object.
+	// Must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character.
```

Comment:
Should update this to match the others I suggested, including the at most 253 characters part